Batched Packet Sending

dr7ana opened the pull request:
- packets batched in groups of 8 to reduce syscall overhead
- packet send logic moved to Endpoint from Connection

dr7ana committed:
Implemented packet batching:
- packets batched in groups of 8 to reduce syscall overhead
- packet send logic moved to Endpoint from Connection

jagerman commented on include/quic/connection.hpp:
```diff
@@ -34,8 +34,10 @@ namespace oxen::quic
 
-        std::array<std::byte, NGTCP2_MAX_UDP_PAYLOAD_SIZE> send_buffer{};
+        static constexpr size_t batch_size = 8;
+        std::array<std::pair<std::array<std::byte, NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE>, size_t>, batch_size> send_buffer{};
```
This type is unwieldy enough (and used in multiple places) that it ought to be a `using send_buffer_t = std::array<...` typedef.

jagerman commented on src/connection.cpp:
```diff
@@ -488,12 +487,16 @@ namespace oxen::quic
+            buf.second = nwrite;
+
+            if (++n_packets == batch_size)
+                break;
```
This doesn't seem right: this `for(;;)` loop may want to send more than 1 packet, but this loop could be restricted to just one iteration (if the previous loop filled up 7/8 slots).  I think we need the same append-and-send-if-full logic as in the upper loop.

jagerman committed:
Allocate buffer to store packets before feeding them into libuv

libuv requires us to keep the buffers alive until it actually does the
send, which doesn't necessarily happen right away.  This means two
things:

- the `uv_udp_send_t` we give it a pointer to has to stay alive until
  it invokes its callback
- we can't give it pointers into our array buffers directly, because we
  may well rewrite those shortly after the call (i.e. as we built more
  packets), so we have to copy all the buffers into allocated storage to
  be deleted after libuv finishes sending.

tewinget commented on src/endpoint.cpp:
```diff
@@ -257,18 +257,53 @@ namespace oxen::quic
+        char* packet_data_pos = packet_data->data();
+        for (int i = 0; i < n_pkts; ++i)
+        {
+            assert(bufs[i].second > 0);
```
```suggestion
            assert(buf[i].second > 0);
```

tewinget commented:
test 004 aborts and 005 fails, but both pass just fine if I rebase my pending `thread_safety` branch on top of this so apart from the suggested change in `endpoint.cpp` I'd say this is good to go.

dr7ana committed:
Merge branch 'oxen-io:dev' into packet_batching

jagerman commented:
Closing (because this is embedded into #1, which takes it much further for performance optimizations)
